wyvern/components/ranking_pipeline.py

```diff
@@ -0,0 +1,176 @@
+# -*- coding: utf-8 -*-
+from typing import Any, Generic, List, Optional
+
+from pydantic import BaseModel
+
+from wyvern.components.business_logic.business_logic import (
+    BusinessLogicPipeline,
+    BusinessLogicRequest,
+)
+from wyvern.components.candidates.candidate_logger import CandidateEventLoggingComponent
+from wyvern.components.events.events import LoggedEvent
+from wyvern.components.impressions.impression_logger import (
+    ImpressionEventLoggingComponent,
+    ImpressionEventLoggingRequest,
+)
+from wyvern.components.models.model_component import ModelComponent, ModelInput
+from wyvern.components.pagination.pagination_component import (
+    PaginationComponent,
+    PaginationRequest,
+)
+from wyvern.components.pagination.pagination_fields import PaginationFields
+from wyvern.components.pipeline_component import PipelineComponent
+from wyvern.entities.candidate_entities import ScoredCandidate
+from wyvern.entities.identifier_entities import QueryEntity
+from wyvern.entities.request import BaseWyvernRequest
+from wyvern.event_logging import event_logger
+from wyvern.wyvern_typing import WYVERN_ENTITY
+
+
+class RankingRequest(
+    BaseWyvernRequest,
+    PaginationFields,
+    Generic[WYVERN_ENTITY],
+):
+    query: QueryEntity
+    candidates: List[WYVERN_ENTITY]
+
+
+class ResponseCandidate(BaseModel):
+    candidate_id: str
+    ranked_score: float
+
+
+class RankingResponse(BaseModel):
+    ranked_candidates: List[ResponseCandidate]
+    events: Optional[List[LoggedEvent[Any]]]
+
+
+class RankingPipeline(
+    PipelineComponent[RankingRequest, RankingResponse],
+    Generic[WYVERN_ENTITY],
+):
+    PATH: str = "/ranking"
+
+    def __init__(self, name: Optional[str] = None):
+        self.pagination_component = PaginationComponent[
+            ScoredCandidate[WYVERN_ENTITY]
+        ]()
+        self.ranking_model = self.model
+        self.candidate_logging_component = CandidateEventLoggingComponent[
+            WYVERN_ENTITY,
+            RankingRequest[WYVERN_ENTITY],
+        ]()
+        self.impression_logging_component = ImpressionEventLoggingComponent[
+            WYVERN_ENTITY,
+            RankingRequest[WYVERN_ENTITY],
+        ]()
+
+        upstream_components = [
+            self.pagination_component,
+            self.ranking_model,
+            self.candidate_logging_component,
+            self.impression_logging_component,
+        ]
+        self.business_logic_pipeline: BusinessLogicPipeline
+        if self.business_logic:
+            self.business_logic_pipeline = self.business_logic
+        else:
+            self.business_logic_pipeline = BusinessLogicPipeline[
+                WYVERN_ENTITY,
+                RankingRequest[WYVERN_ENTITY],
+            ]()
+            upstream_components.append(self.business_logic_pipeline)
+
+        super().__init__(
+            *upstream_components,
+            name=name,
+        )
+
+    @property
```

Comment:
out of curiosity, why should we make the model a property? what do we get out of it?

Reply:
`self.model` vs `self.model()`

I think it makes sense if the function is probably `self.get_model()` (the function name is not a noun)

but you made a great point: having wyvern user to define the `@property` is probably annoying.

but we have already established this kinda "annoying" pattern:
https://github.com/Wyvern-AI/wyvern/blob/main/wyvern/components/component.py#L122-L130

Since we have a plan to add python decorator to help build pipeline/component, let's improve this user experience at that point

Reply:
Ah yeah, I don't think this should have been a `@property`. I actually don't like that annotation since it doesn't cache by default

Reply:
@suchintan @ykeremy 

How about `def get_model(self)`?